aten/src/ATen/native/SegmentReduce.cpp

```diff
@@ -60,6 +61,54 @@ Tensor _segment_reduce_cpu_kernel(
   return output;
 }
 
+Tensor _segment_reduce_cpu_backward_kernel(
+    const Tensor& grad,
+    const Tensor& output,
+    const Tensor& data,
+    const Tensor& lengths) {
+  auto grad_output = at::zeros({data.sizes()}, grad.options());
+
+  int64_t batch_size = lengths.numel();
+  const auto data_contig = data.contiguous();
+
+  const auto lengths_contig = lengths.contiguous();
+  const auto* lengths_data = lengths_contig.data_ptr<int64_t>();
+
+  AT_DISPATCH_ALL_TYPES_AND2(
+      kBFloat16,
+      kHalf,
+      data_contig.scalar_type(),
+      "_segment_reduce_cpu",
+      ([&]() {
+        auto* output_data = output.data_ptr<scalar_t>();
+        auto* grad_output_data = grad_output.data_ptr<scalar_t>();
+        const auto* values_data = data_contig.data_ptr<scalar_t>();
+        int64_t k = 0;
+        for (int64_t i = 0; i < batch_size; ++i) {
+          // scalar_t reduction = std::numeric_limits<scalar_t>::lowest();
+          int64_t counter = 0;
+          for (int64_t j = 0; j < lengths_data[i]; ++j) {
+            if (at::_isnan(values_data[k]) ||
+                values_data[k] == output_data[i]) {
+              grad_output_data[k] = 1;
```

Comment:
`grad_output_data[k] = grad[i]`, and compare `grad_output_data[index] !=0` in the next loop. Also, as pointed above, the name should be `grad_input`.